src/main/java/org/pcollections/OrderedPSet.java

```diff
@@ -33,18 +33,19 @@ public static <E> OrderedPSet<E> singleton(final E e) {
     return OrderedPSet.<E>empty().plus(e);
   }
 
-  private final PSet<E> contents;
-  private final PVector<E> order;
+  private final PMap<E, Long> ids;
+  private final PSortedMap<Long, E> elements;
 
-  private OrderedPSet(final PSet<E> c, final PVector<E> o) {
-    contents = c;
-    order = o;
+  private OrderedPSet(final PMap<E, Long> ids, final PSortedMap<Long, E> elements) {
+    this.ids = ids;
+    this.elements = elements;
   }
 
   @Override
   public OrderedPSet<E> plus(final E e) {
-    if (contents.contains(e)) return this;
-    return new OrderedPSet<E>(contents.plus(e), order.plus(e));
+    if (ids.containsKey(e)) return this;
+    final Long id = elements.isEmpty() ? -Long.MIN_VALUE : (elements.lastKey() + 1);
```

Comment:
`-Long.MIN_VALUE` equals just `Long.MIN_VALUE` due to overflow. But I think you meant `MIN_VALUE` anyway, so I suppose this "bug" is harmless?

Reply:
haha whoops, good find, thanks! I'm glad it still works but yeah that was definitely not what I intended